docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1438 @@
+}
+
+type GangSchedulerPlugin string
+
+const (
+    GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"
+    GangSchedulerPlugins       GangSchedulerPlugin = "scheduler-plugins"
+)
+```
+
+### The Torch Spec API
+
+The `TorchSpec` API represents the configuration for the PyTorch distributed training. This configuration
+allows platform engineers to explicitly configure `torchrun` setting.
+
+The distributed parameters are taken from the
+[PyTorch distributed launch run](https://github.com/pytorch/pytorch/blob/94dc3253a0fefbfb95fbe467ddd628e4c2eb08d7/torch/distributed/run.py).
+
+For Elastic Training we will always pass the following parameters:
+
+```bash
+--rdzv-backend=c10d
+
+--rdzv-id will be set automatically.
+
+--rdzv-endpoint will always point to the node-0 Pod.
+```
+
+Since the [etcd and etcd-v2 are legacy rendezvous](https://pytorch.org/docs/stable/elastic/run.html#note-on-rendezvous-backend),
+we won't support them in `TorchSpec`. We can introduce them in the future if users will require them.
+
+```golang
+// TorchSpec represents the configuration for PyTorch.
+type TorchSpec struct {
+
+    // Number of Procs per Node.
+    NumProcPerNode int `json:"numProcPerNode,omitempty"`
+
+    // Used for single-node multi-worker training
+    Standalone bool `json:"standalone,omitempty"`
+
+    // Torch Elastic Policy.
+    ElasticPolicy *TorchElasticPolicy `json:"elasticPolicy,omitempty"`
+}
+
+// If the Elastic Policy is set, the numNodes parameter is ignored.
+// --nnodes=minNodes:maxNodes
+type TorchElasticPolicy struct {
+
+    // The limits to restart TrainJob.
+    // Insert it to the JobSet.spec.failurePolicy.maxRestarts
+    MaxRestarts *in32 `json:"maxRestarts,omitempty"`
+
+    // Min number of nodes for HPA and torchrun.
+    MinNodes *in32 `json:"minNodes,omitempty"`
+
+    // Max number of nodes for HPA and torchrun.
+    MaxNodes *in32 `json:"maxNodes,omitempty"`
+
+    // Metrics for scale up and down replicas.
+    Metrics []autoscalingv2.MetricSpec `json:"metrics,omitempty"`
+}
+```
+
+### The MPI Spec API
+
+The `MPISpec` API represents the configuration for training using MPI orchestration.
+E.g. creation of host-files and SSH keys. Using MPI might be more efficient for training on HPC
+clusters or for some ML frameworks (e.g. [MLX distributed with MPI](https://ml-explore.github.io/mlx/build/html/usage/distributed.html)).
+
+We will fully migrate to the MPI Operator V2 functionality as part of this KEP.
+Check [the proposal for the MPI V2 APIs.](https://github.com/kubeflow/mpi-operator/blob/master/proposals/scalable-robust-operator.md)
+
+```golang
+type MPISpec struct {
+    // Number of Procs per Node.
+    NumProcPerNode int `json:"numProcPerNode,omitempty"`
+
+    // MPI Implementation to create appropriate host-files.
+    // Can be one of OpenMPI, Intel, or MPICH.
+    MPIImplementation MPIImplementation `json:"mpiImplementation,omitempty"`
```

Comment:
```suggestion
    MPIImplementation MPIImplementation `json:"mpiImplementation"`
```

Let's use this field as a necessary field and then default to OpenMPI when users do not specify this field.